### PR TITLE
[shopsys] moved Persoo feeds to Luigi's Box feeds

### PR DESCRIPTION
**Pull request by TomasLudvik**
| Q             | A
| ------------- | ---
|Description, the reason for the PR| We decided before finishing of Persoo to migrate to Luigi's Box.
|New feature| No 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

----
:globe_with_meridians: Live Preview:
  - https://tl-move-persoo-feeds-to-luigis-box.odin.shopsys.cloud
  - https://cz.tl-move-persoo-feeds-to-luigis-box.odin.shopsys.cloud

### CODE REVIEW

File: UPGRADE-14.0.md [hunk 1]
```diff
@@ -524,7 +524,7 @@ Follow the instructions in relevant sections, e.g. `shopsys/coding-standards` or
     -   see #project-base-diff to update your project
 -   remove preorder and vendor delivery date features ([#2942](https://github.com/shopsys/shopsys/pull/2942))
     -   see #project-base-diff to update your project
--   add Persoo Product feed and remove unused code ([#2939](https://github.com/shopsys/shopsys/pull/2939))
+-   add Luigi's Box Product feed and remove unused code ([#2939](https://github.com/shopsys/shopsys/pull/2939))
```
Comment (vitek-rostislav) [inline UPGRADE-14.0.md:527]: the links in the upgrade notes might be a little confusing now (the line says "Luigis box" but the PR is about Persoo) but what can we do :shrug:

File: composer.json [hunk 2]
```diff
@@ -266,7 +267,7 @@
         "shopsys/product-feed-google": "self.version",
         "shopsys/product-feed-zbozi": "self.version",
         "shopsys/product-feed-heureka": "self.version",
-        "shopsys/product-feed-persoo": "self.version",
+        "shopsys/product-feed-luigis-box": "self.version",
```
Comment (vitek-rostislav) [inline composer.json:270]: is it possible to rename the packages or will we rather create new repositories?

File: packages/article-feed-luigis-box/src/Model/LuigisBoxArticleFeedItem.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+{
+    protected const UNIQUE_IDENTIFIER_PREFIX = 'article-';
+
+    /**
+     * @param int $id
+     * @param string $title
+     * @param string $link
+     * @param string|null $description
+     */
+    public function __construct(
```
Comment (vitek-rostislav) [inline packages/article-feed-luigis-box/src/Model/LuigisBoxArticleFeedItem.php:19]: why there is no image link anymore?

File: packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFacade.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        protected readonly LuigisBoxCategoryRepository $luigisBoxCategoryRepository,
+        protected readonly LuigisBoxCategoryFeedItemFactory $luigisBoxFeedItemFactory,
+    ) {
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Domain\Config\DomainConfig $domainConfig
+     * @param int|null $lastSeekId
+     * @param int $maxResults
+     * @return \Shopsys\CategoryFeed\LuigisBoxBundle\Model\FeedItem\LuigisBoxCategoryFeedItem[]|iterable
```
Comment (vitek-rostislav) [inline packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFacade.php]: I believe the proper annotation is:
```suggestion
     * @return iterable<int, \Shopsys\CategoryFeed\LuigisBoxBundle\Model\FeedItem\LuigisBoxCategoryFeedItem>
```

File: packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFactory.php [hunk 5]
```diff
@@ -30,35 +29,31 @@ public function __construct(
...
         }
 
-        return new PersooCategoryFeedItem(
+        return new LuigisBoxCategoryFeedItem(
             $category->getId(),
             $category->getName($locale),
-            array_reverse($hierarchyIds),
             $this->friendlyUrlFacade->getAbsoluteUrlByRouteNameAndEntityId($domainConfig->getId(), 'front_product_list', $category->getId()),
             array_reverse($hierarchyNames),
-            TransformString::convertHtmlToPlainText($category->getDescription($domainConfig->getId())),
```
Comment (vitek-rostislav) [inline packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFactory.php:61]: missing category description is also quite sad, they do not support it?

File: packages/category-feed-luigis-box/src/Resources/views/feed.xml.twig [hunk 6]
```diff
@@ -0,0 +1,18 @@
+{% block begin %}
+<?xml version="1.0" encoding="utf-8"?>
+<categories>
+{% endblock %}
+
+{% block item %}
+    <category>
+        <identity>{{ item.identity}}</identity>
```
Comment (vitek-rostislav) [inline packages/category-feed-luigis-box/src/Resources/views/feed.xml.twig]:
```suggestion
        <identity>{{ item.identity }}</identity>
```

nitpick :smile:

File: packages/framework/src/Component/Image/ImageUrlWithSizeHelper.php [hunk 7]
```diff
@@ -2,19 +2,20 @@
...
      * @param string $imageUrl
      * @param int|null $width
+     * @param int|null $height
      * @return string
      */
-    public static function limitWidthInImageUrl(string $imageUrl, ?int $width = null): string
+    public static function limitWidthInImageUrl(string $imageUrl, ?int $width = null, ?int $height = null): string
     {
-        return $imageUrl . '?width=' . ($width ?? static::DEFAULT_IMAGE_SIZE);
+        return $imageUrl . '?width=' . ($width ?? static::DEFAULT_IMAGE_SIZE) . ($height !== null ? '&height=' . $height : '');
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Image/ImageUrlWithSizeHelper.php:19]: I am not sure whether this would work as expected - there is no cropping so only one dimension is taken into account and the image is resized based on that (e.g. https://14-0.odin.shopsys.cloud/content/images/product/22-sencor-sle-22f46dm4-hello-kitty_1.jpg?width=605&height=605 does not result into a square image).

Moreover, the method name does not respect the behavior anymore (it only mentions width while there is also a `height` parameter). Let's discuss it, please

Comment (vitek-rostislav) [reply packages/framework/src/Component/Image/ImageUrlWithSizeHelper.php:19]: btw, the cropping can be achieved like this:
https://14-0.odin.shopsys.cloud/content/images/product/22-sencor-sle-22f46dm4-hello-kitty_1.jpg?width=605&height=605&resize=fill

File: packages/product-feed-luigis-box/src/Model/FeedItem/LuigisBoxFeedItem.php [hunk 8]
```diff
@@ -99,81 +107,121 @@ public function getDescription(): ?string
...
+    public function getAvailabilityRankText(): string
+    {
+        return $this->availabilityText;
     }
 
     /**
      * @return int
      */
-    public function getCategoryId(): int
+    public function getAvailabilityRank(): int
```
Comment (vitek-rostislav) [inline packages/product-feed-luigis-box/src/Model/FeedItem/LuigisBoxFeedItem.php]: there are quite a lot of magic numbers in this method, could you please extract them into constants?

Comment (vitek-rostislav) [reply packages/product-feed-luigis-box/src/Model/FeedItem/LuigisBoxFeedItem.php]: btw, it is funny that `availableInDays` now can be only `0` or `null` :slightly_smiling_face: 
(based on `ProductAvailabilityFacade::getProductAvailabilityDaysByDomainId` implementation)

File: packages/product-feed-luigis-box/src/Model/Product/LuigisBoxProductRepository.php [hunk 9]
```diff
@@ -31,8 +32,12 @@ public function getProducts(
...
     ): iterable {
         $queryBuilder = $this->productRepository->getAllOfferedQueryBuilder($domainConfig->getId(), $pricingGroup)
-            ->addSelect('b')->leftJoin('p.brand', 'b')
-            ->orderBy('p.id', 'asc')
+            ->addSelect('b')
+            ->leftJoin('p.brand', 'b')
+            ->andWhere('p.variantType != :mainVariantType')
+            ->setParameter('mainVariantType', Product::VARIANT_TYPE_MAIN)
+            ->leftJoin('p.mainVariant', 'mv')
+            ->orderBy('mv.id', 'asc')
```
Comment (vitek-rostislav) [inline packages/product-feed-luigis-box/src/Model/Product/LuigisBoxProductRepository.php]: Why is this necessary, please? 
Could not be the same result achieved without the join just by `->orderBy('IDENTITY(p.mainVariant)', 'asc')`? (I am not sure whether this would work :smile: )

Comment (vitek-rostislav) [reply packages/product-feed-luigis-box/src/Model/Product/LuigisBoxProductRepository.php]: ah, I have found a similar usage in our code - https://github.com/shopsys/shopsys/blob/14.0/project-base/app/src/Model/Product/Parameter/ParameterRepository.php#L330

File: packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            <category{% if categoryId is same as(item.mainCategoryId) %} primary="true"{% endif %}>{{ categoryNames }}</category>
+        {% endfor %}
+        {%  if item.brand is not null -%}
+            <brand>{{ item.brand }}</brand>
+        {% endif -%}
+        <price>{{ item.price.priceWithVat|priceWithCurrencyByCurrencyId(item.currency.id) }}</price>
+        {% if item.imageLinkS is not null -%}
+            <image_link_s>{{ item.imageLinkS }}</image_link_s>
+            <image_link_m>{{ item.imageLinkM }}</image_link_m>
+            <image_link_l>{{ item.imageLinkL }}</image_link_l>
```
Comment (vitek-rostislav) [inline packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig:25]: it feels like two `if` conditions are missing here, right?

Comment (vitek-rostislav) [reply packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig:25]: ok, maybe they are not necessary - if the `imageLinkS` is not null, then the other two are not null either, but it is a little bit confusing IMHO :innocent:

File: packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        {% endif -%}
+        <description>{{ item.description }}</description>
```
Comment (vitek-rostislav) [inline packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig]: the description is nullable, should not this be wrapped in the `if` condition then?

File: packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        {% if item.flagNames|length > 0 -%}
+            <labels>{{ item.flagNames|join(' | ') }}</labels>
+        {% endif %}
+        <product_code>{{ item.productCode }}</product_code>
+        <ean>{{ item.ean }}</ean>
```
Comment (vitek-rostislav) [inline packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig]: also nullable properties

File: packages/product-feed-luigis-box/tests/Unit/LuigisBoxFeedItemTest.php [hunk 13]
```diff
@@ -199,75 +200,74 @@ private function mockProductUrl(Product $product, DomainConfig $domain, string $
...
+        self::assertInstanceOf(LuigisBoxFeedItem::class, $luigisBoxFeedItem);
+        self::assertEquals(self::PRODUCT_IDENTITY, $luigisBoxFeedItem->getIdentity());
+        self::assertEquals(self::PRODUCT_ID, $luigisBoxFeedItem->getSeekId());
+        self::assertEquals(self::PRODUCT_NAME, $luigisBoxFeedItem->getTitle());
+        self::assertEquals(self::PRODUCT_URL, $luigisBoxFeedItem->getUrl());
+        self::assertEquals(self::PRODUCT_EAN, $luigisBoxFeedItem->getEan());
+        self::assertEquals(self::PRODUCT_PART_NO, $luigisBoxFeedItem->getProductCode());
+        self::assertEquals(self::PRODUCT_SKU, $luigisBoxFeedItem->getSku());
+        self::assertNull($luigisBoxFeedItem->getDescription());
+        self::assertEquals([0 => self::MAIN_CATEGORY_NAME], $luigisBoxFeedItem->getCategoryNamesIndexedByCategoryId());
```
Comment (vitek-rostislav) [inline packages/product-feed-luigis-box/tests/Unit/LuigisBoxFeedItemTest.php]: I would expect `[self::MAIN_CATEGORY_ID => self::MAIN_CATEGORY_NAME]` here, how come?

**Review by vitek-rostislav — CHANGES_REQUESTED** [review]
Hi, nice job there, it must have been exhausting to re-work everything :pray: 
I have added a few notes so please check them out. We can discuss it personally :wink:

File: packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFacade.php (hunk 4 above)
Comment (vitek-rostislav) [reply packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFacade.php]: (similarly in `LuigisBoxFeedItemFacade::getItems`)

btw, I do not want to be that guy, but IMHO the navigation is a little bit confusing with the naming:
- `LuigisBoxFeedItemFacade`
- `LuigisBoxCategoryFeedItemFacade`
- `LuigisBoxArticleFacade`

it is not obvious that the first one is for products (the same applies to `LugisBoxFeedItem`...and yes, it is the name of the bundle, but...) and the naming is not unified so I had a hard time finding the proper files by their names.

File: UPGRADE-14.0.md (hunk 1 above)
Comment (TomasLudvik) [reply UPGRADE-14.0.md:527]: Yes :)

File: composer.json (hunk 2 above)
Comment (TomasLudvik) [reply composer.json:270]: I will do this with merge

File: packages/article-feed-luigis-box/src/Model/LuigisBoxArticleFeedItem.php (hunk 3 above)
Comment (TomasLudvik) [reply packages/article-feed-luigis-box/src/Model/LuigisBoxArticleFeedItem.php:19]: Because there is no image in Luigi's Box feed

File: packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFactory.php (hunk 5 above)
Comment (TomasLudvik) [reply packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFactory.php:61]: Exactly

File: packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFacade.php (hunk 4 above)
Comment (TomasLudvik) [reply packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFacade.php]: I have introduced some naming changes to unify it

File: packages/category-feed-luigis-box/src/Resources/views/feed.xml.twig (hunk 6 above)
Comment (TomasLudvik) [reply packages/category-feed-luigis-box/src/Resources/views/feed.xml.twig]: fixed

File: packages/framework/src/Component/Image/ImageUrlWithSizeHelper.php (hunk 7 above)
Comment (TomasLudvik) [reply packages/framework/src/Component/Image/ImageUrlWithSizeHelper.php:19]: I have renamed it

Also, I think there is no need to introduce cropping, just limiting size is enough

File: packages/product-feed-luigis-box/src/Model/FeedItem/LuigisBoxFeedItem.php (hunk 8 above)
Comment (TomasLudvik) [reply packages/product-feed-luigis-box/src/Model/FeedItem/LuigisBoxFeedItem.php]: I have added some constants. This functionality is very improvised as we have nothing of such kind implemented.

File: packages/product-feed-luigis-box/src/Model/Product/LuigisBoxProductRepository.php (hunk 9 above)
Comment (TomasLudvik) [reply packages/product-feed-luigis-box/src/Model/Product/LuigisBoxProductRepository.php]: changed

File: packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig (hunk 10 above)
Comment (TomasLudvik) [reply packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig:25]: I do not think that adding two redundant conditions would improve it

File: packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig (hunk 11 above)
Comment (TomasLudvik) [reply packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig]: Added

File: packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig (hunk 12 above)
Comment (TomasLudvik) [reply packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig]: Added

File: packages/product-feed-luigis-box/tests/Unit/LuigisBoxFeedItemTest.php (hunk 13 above)
Comment (TomasLudvik) [reply packages/product-feed-luigis-box/tests/Unit/LuigisBoxFeedItemTest.php]: Fixed

File: packages/article-feed-luigis-box/src/Model/LuigisBoxArticleFeedItem.php (hunk 3 above)
Comment (vitek-rostislav) [reply packages/article-feed-luigis-box/src/Model/LuigisBoxArticleFeedItem.php:19]: well that is dissapointing

File: packages/framework/src/Component/Image/ImageUrlWithSizeHelper.php (hunk 7 above)
Comment (vitek-rostislav) [reply packages/framework/src/Component/Image/ImageUrlWithSizeHelper.php:19]: ok

File: packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig (hunk 10 above)
Comment (vitek-rostislav) [reply packages/product-feed-luigis-box/src/Resources/views/feed.xml.twig:25]: well it would be definitely less confusing but I do not insist on that

File: packages/product-feed-luigis-box/src/Model/FeedItem/LuigisBoxFeedItem.php (hunk 8 above)
Comment (vitek-rostislav) [reply packages/product-feed-luigis-box/src/Model/FeedItem/LuigisBoxFeedItem.php]: ok, thanks for the fix and info :slightly_smiling_face:

File: packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFacade.php (hunk 4 above)
Comment (vitek-rostislav) [reply packages/category-feed-luigis-box/src/Model/FeedItem/LuigisBoxCategoryFeedItemFacade.php]: thank you, now it is much better :pray: 

Two more things:
- could you please fix the iterable annotation in `LuigisBoxProductFeedItemFacade::getItems()`?
- after the classes renaming, some of the property and variable names do not match but I believe I can live with that and if you do not want to fix it, I understand :blush:

**Review by vitek-rostislav — APPROVED** [review]
